server/jdkhttp-server/src/main/scala/sttp/tapir/server/jdkhttp/internal/ParsedMultiPart.scala

```diff
@@ -33,59 +51,142 @@ case class ParsedMultiPart(headers: Map[String, Seq[String]], body: Array[Byte])
+    private val buffer = new mutable.ArrayBuffer[Byte](bufferSize)
+    private var done = false
+    private var currentPart: ParsedMultiPart = new ParsedMultiPart()
+    private var bodySize: Int = 0
+    private var stream: PartStream = ByteStream()
+    private var parseState: ParseStatus = LookForInitialBoundary
+
+    private val initialBoundaryMatcher = new KMPMatcher(boundary ++ CRLF)
+    private val bodyBoundaryMatcher = new KMPMatcher(CRLF ++ boundary ++ CRLF)
+    private val endMatcher = new KMPMatcher(CRLF ++ boundary ++ END_DELIMITER)
+
+    def isDone: Boolean = done
+
+    def updateStateWith(currentByte: Int): Unit = {
+      buffer.addOne(currentByte.toByte)
+      this.parseState match {
+        case LookForInitialBoundary => parseInitialBoundary(currentByte)
+        case ParsePartHeaders       => parsePartHeaders()
+        case ParsePartBody          => parsePartBody(currentByte)
+      }
+    }
+
+    private def parseInitialBoundary(currentByte: Int): Unit = {
+      val foundBoundary = initialBoundaryMatcher.matchByte(currentByte.toByte)
+      if (foundBoundary) {
+        changeState(ParsePartHeaders)
+      }
+    }
+
+    private def parsePartHeaders(): Unit = {
+      if (buffer.endsWith(CRLF)) {
+        buffer.view.map(_.toChar).mkString match {
+          case headerLine if !headerLine.isBlank => addHeader(headerLine)
+          case _                                 => changeState(ParsePartBody)
+        }
+        buffer.clear()
+      } else if (buffer.length >= bufferSize) {
+        throw new RuntimeException(
+          s"Reached max size of $bufferSize bytes before reaching newline when parsing header."
         )
-      case None => changeState(AfterBoundary)
+      }
     }
-  }
 
-  def parseMultipartBody(httpExchange: HttpExchange, boundary: String): Seq[ParsedMultiPart] = {
-    val reader = new BufferedReader(new InputStreamReader(httpExchange.getRequestBody))
-    val initialParseState: ParseData = ParseData(empty, List.empty, Default)
-    Iterator
-      .continually(reader.readLine())
-      .takeWhile(_ != null)
-      .foldLeft(initialParseState) { case (state, line) =>
-        state.parseState match {
-          case Default if line.startsWith(boundary)           => state.changeState(AfterBoundary)
-          case Default                                        => state
-          case AfterBoundary if line.trim.isEmpty             => state.changeState(AfterHeaderSpace)
-          case AfterBoundary                                  => state.addHeader(line)
-          case AfterHeaderSpace if !line.startsWith(boundary) => state.addBody(line.getBytes())
-          case AfterHeaderSpace                               => state.completePart()
+    private def parsePartBody(currentByte: Int): Unit = {
+      bodySize += 1
+      val foundFinalBoundary = endMatcher.matchByte(currentByte.toByte)
+      val foundBoundary = bodyBoundaryMatcher.matchByte(currentByte.toByte) || foundFinalBoundary
```

Comment:
I guess this could be done with another state that looks for `--` / `CRLF` instead.
Not sure which one i prefer. I guess adding the state would be marginally more performant. What do you think?

Reply:
I think that's good as-is :)